src/dictionary.c

```diff
@@ -535,42 +536,47 @@ SEXP vctrs_count(SEXP x) {
   struct dictionary* d = new_dictionary(x);
   PROTECT_DICT(d, &nprot);
 
-  SEXP val = PROTECT_N(Rf_allocVector(INTSXP, d->size), &nprot);
-  int* p_val = INTEGER(val);
+  SEXP count = PROTECT_N(Rf_allocVector(INTSXP, d->size), &nprot);
+  int* p_count = INTEGER(count);
 
   for (int i = 0; i < n; ++i) {
     uint32_t hash = dict_hash_scalar(d, i);
 
     if (d->key[hash] == DICT_EMPTY) {
       dict_put(d, hash, i);
-      p_val[hash] = 0;
+      p_count[hash] = 0;
     }
-    p_val[hash]++;
+    p_count[hash]++;
   }
 
   // Create output
-  SEXP out_key = PROTECT_N(Rf_allocVector(INTSXP, d->used), &nprot);
-  SEXP out_val = PROTECT_N(Rf_allocVector(INTSXP, d->used), &nprot);
-  int* p_out_key = INTEGER(out_key);
-  int* p_out_val = INTEGER(out_val);
+  SEXP out_loc = PROTECT_N(Rf_allocVector(INTSXP, d->used), &nprot);
+  int* p_out_loc = INTEGER(out_loc);
+
+  // Reuse `count` storage, which will be narrowed
+  SEXP out_count = count;
+  int* p_out_count = p_count;
```

Comment:
I was able to optimize this while I was here to avoid an allocation of size `d->used`

We can reuse the storage of `count` for the output `out_count`, and just resize it at the end